modules/google-analytics/wp-google-analytics.php

```diff
@@ -0,0 +1,248 @@
+ * It helps us avoid name collisions
+ * http://codex.wordpress.org/Writing_a_Plugin#Avoiding_Function_Name_Collisions
+ */
+class JetpackGoogleAnalytics {
+
+	/**
+	 * @var JetpackGoogleAnalytics - Static property to hold our singleton instance
+	 */
+	static $instance = false;
+
+	var $tokens = array();
+
+	/**
+	 * This is our constructor, which is private to force the use of get_instance()
+	 *
+	 * @return void
+	 */
+	private function __construct() {
+		add_action( 'get_footer',               array( $this, 'insert_code' ) );
+		add_action( 'wp_enqueue_scripts',       array( $this, 'track_outgoing' ) );
+	}
+
+	/**
+	 * Function to instantiate our class and make it a singleton
+	 */
+	public static function get_instance() {
+		if ( ! self::$instance ) {
+			self::$instance = new self;
+		}
+
+		return self::$instance;
+	}
+
+	/**
+	 * Used to generate a tracking URL
+	 *
+	 * @param array $track - Must have ['data'] and ['code'].
+	 * @return string - Tracking URL
+	 */
+	private function _get_url( $track ) {
+		$site_url = ( is_ssl() ? 'https://':'http://' ) . sanitize_text_field( wp_unslash( $_SERVER['HTTP_HOST'] ) ); // Input var okay.
+		foreach ( $track as $k => $value ) {
+			if ( strpos( strtolower( $value ), strtolower( $site_url ) ) === 0 ) {
+				$track[ $k ] = substr( $track[ $k ], strlen( $site_url ) );
+			}
+			if ( 'data' === $k ) {
+				$track[ $k ] = preg_replace( '/^https?:\/\/|^\/+/i', '', $track[ $k ] );
+			}
+
+			// This way we don't lose search data.
+			if ( 'data' === $k && 'search' === $track['code'] ) {
+				$track[ $k ] = rawurlencode( $track[ $k ] );
+			} else {
+				$track[ $k ] = preg_replace( '/[^a-z0-9\.\/\+\?=-]+/i', '_', $track[ $k ] );
+			}
+
+			$track[ $k ] = trim( $track[ $k ], '_' );
+		}
+		$char = ( strpos( $track['data'], '?' ) === false ) ? '?' : '&amp;';
+		return str_replace( "'", "\'", "/{$track['code']}/{$track['data']}{$char}referer=" . rawurlencode( isset( $_SERVER['HTTP_REFERER'] ) ? $_SERVER['HTTP_REFERER'] : '' ) ); // Input var okay.
+	}
+
+	/**
+	 * Maybe output or return, depending on the context
+	 */
+	private function _output_or_return( $val, $maybe ) {
+		if ( $maybe ) {
+			echo $val . "\r\n";
+		} else {
+			return $val;
+		}
+	}
+
+	/**
+	 * This injects the Google Analytics code into the footer of the page.
+	 *
+	 * @param bool[optional] $output - defaults to true, false returns but does NOT echo the code.
+	 */
+	public function insert_code( $output = true ) {
+		// If $output is not a boolean false, set it to true (default).
+		$output = ( false !== $output);
+
+		$tracking_id = $this->_get_options( 'code' );
+		if ( empty( $tracking_id ) ) {
+			return $this->_output_or_return( '<!-- Your Google Analytics Plugin is missing the tracking ID -->', $output );
+		}
+
+		// get our plugin options.
+		$wga = $this->_get_options();
+		// If the user's role has wga_no_track set to true, return without inserting code.
```

Comment:
We should be to simplify a lot of this function, since most of these options can't be set by the site owners